File: orange_inventory/orange_inventory/page/solution_assets/solution_assets.js
frappe.pages["solution-assets"].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Активи за Рішенням",
		single_column: true,
	});

	wrapper.solution_assets_page = new SolutionAssetsPage(page);
};

frappe.pages["solution-assets"].on_page_show = function (wrapper) {
	// Перевіряємо, чи є рішення в URL
	const urlParams = new URLSearchParams(window.location.search);
	const decision = urlParams.get("decision");
	if (decision && wrapper.solution_assets_page) {
		wrapper.solution_assets_page.decision_field.set_value(decision);
	}
};

class SolutionAssetsPage {
	constructor(page) {
		this.page = page;
		this.make();
	}

	make() {
		this.make_filters();
		this.make_results_area();
	}

	make_filters() {
		// Поле вибору рішення
		this.decision_field = this.page.add_field({
			fieldname: "decision",
			label: __("Рішення"),
			fieldtype: "Link",
			options: "oiBasisDoc",
			change: () => {
				this.load_assets();
			},
		});
	}

	make_results_area() {
		// Контейнер для результатів
		this.results_wrapper = $(`
			<div class="solution-assets-results" style="margin-top: 20px;">
				<div class="text-muted text-center" style="padding: 40px;">
					${__("Виберіть рішення для перегляду активів")}
				</div>
			</div>
		`).appendTo(this.page.body);
	}

	load_assets() {
		const decision = this.decision_field.get_value();
		if (!decision) {
			this.results_wrapper.html(`
				<div class="text-muted text-center" style="padding: 40px;">
					${__("Виберіть рішення для перегляду активів")}
				</div>
			`);
			return;
		}

		this.results_wrapper.html(`
			<div class="text-center" style="padding: 40px;">
				<div class="spinner-border text-primary" role="status"></div>
				<div class="mt-2">${__("Завантаження...")}</div>
			</div>
		`);

		frappe.call({
			method: "orange_inventory.orange_inventory.page.solution_assets.solution_assets.get_assets_by_decision",
			args: { decision: decision },
			callback: (r) => {
				if (r.message) {
					this.render_results(r.message);
				}
			},
			error: () => {
				this.results_wrapper.html(`
					<div class="text-danger text-center" style="padding: 40px;">
						${__("Помилка завантаження даних")}
					</div>
				`);
			},
		});
	}

	render_results(data) {
		if (!data.assets || data.assets.length === 0) {
			this.results_wrapper.html(`
				<div class="text-muted text-center" style="padding: 40px;">
					${__("Немає активів за цим рішенням")}
				</div>
			`);
			return;
		}

		// Групування активів по власнику
		const grouped = this.group_by_owner(data.assets);

		// Інформація про рішення
		let decision_info = `
			<div class="card mb-4">
				<div class="card-header">
					<strong>${__("Рішення")}: ${data.decision_info.full_title || data.decision_info.title}</strong>
				</div>
				<div class="card-body">
					<div class="row">
						<div class="col-md-4">
							<strong>${__("Номер")}:</strong> ${data.decision_info.decision_number}
						</div>
						<div class="col-md-4">
							<strong>${__("Дата")}:</strong> ${frappe.datetime.str_to_user(data.decision_info.decision_date)}
						</div>
						<div class="col-md-4">
							<strong>${__("Всього активів")}:</strong> ${data.assets.length}
						</div>
					</div>
				</div>
			</div>
		`;

		// Таблиці активів згруповані по власнику
		let assets_tables = "";
		for (const [ownerKey, ownerData] of Object.entries(grouped)) {
			const ownerName = ownerData.owner_name || __("Без власника");
			const ownerLink = ownerData.owner_id
				? `<a href="/app/oiorganization/${ownerData.owner_id}">${ownerName}</a>`
				: ownerName;

			assets_tables += `
				<div class="card mb-4">
					<div class="card-header d-flex justify-content-between align-items-center">
						<div>
							<strong>${ownerLink}</strong>
							<span class="text-muted ml-2">(${ownerData.assets.length} ${__("активів")})</span>
						</div>
						<div>
							<span class="badge badge-primary">${__("Кількість")}: ${ownerData.total_quantity}</span>
							<span class="badge badge-success ml-1">${__("Вартість")}: ${frappe.format(ownerData.total_value, {
				fieldtype: "Currency",
			})}</span>
						</div>
					</div>
					<div class="card-body p-0">
						<table class="table table-bordered table-hover mb-0">
							<thead class="thead-light">
								<tr>
									<th style="width: 30px;"></th>
									<th>${__("Назва")}</th>
									<th>${__("Серійний номер")}</th>
									<th>${__("Інвентарний номер")}</th>
									<th>${__("Статус")}</th>
									<th>${__("Кількість")}</th>
									<th>${__("Вартість")}</th>
								</tr>
							</thead>
							<tbody>
								${ownerData.assets
									.map(
										(asset) => `
									<tr class="asset-row" data-asset="${asset.name}">
										<td class="text-center">
											<button class="btn btn-xs btn-default toggle-history" data-asset="${asset.name}" title="${__(
											"Історія передачі"
										)}">
												<i class="fa fa-chevron-right"></i>
											</button>
										</td>
										<td>
											<a href="/app/oiasset/${asset.name}">${asset.asset_name}</a>
										</td>
										<td>${asset.serial_no || "-"}</td>
										<td>${asset.inventory_no || "-"}</td>
										<td>
											<span class="badge ${this.get_status_class(asset.status)}">${asset.status}</span>
										</td>
										<td class="text-right">${asset.quantity}</td>
										<td class="text-right">${frappe.format(asset.total, { fieldtype: "Currency" })}</td>
									</tr>
									<tr class="history-row d-none" data-asset="${asset.name}">
										<td colspan="7" class="p-0">
											<div class="history-content p-3 bg-light">
												${this.render_movement_history(asset.movement_history)}
											</div>
										</td>
									</tr>
								`
									)
									.join("")}
							</tbody>
						</table>
					</div>
				</div>
			`;
		}

		// Загальний підсумок
		let summary = `
			<div class="card mt-4">
				<div class="card-header">
					<strong>${__("Загальний підсумок")}</strong>
				</div>
				<div class="card-body">
					<div class="row">
						<div class="col-md-3">
							<strong>${__("Загальна кількість")}:</strong> ${data.summary.total_quantity}
						</div>
						<div class="col-md-3">
							<strong>${__("Загальна вартість")}:</strong> ${frappe.format(data.summary.total_value, {
			fieldtype: "Currency",
		})}
						</div>
						<div class="col-md-3">
							<strong>${__("Власників")}:</strong> ${Object.keys(grouped).length}
						</div>
						<div class="col-md-3">
							<strong>${__("Статуси")}:</strong>
							${Object.entries(data.summary.status_counts)
								.map(
									([status, count]) =>
										`<span class="badge ${this.get_status_class(
											status
										)} mr-1">${status}: ${count}</span>`
								)
								.join("")}
						</div>
					</div>
				</div>
			</div>
		`;

		this.results_wrapper.html(decision_info + assets_tables + summary);

		// Обробник кліку для розгортання історії
		this.results_wrapper.find(".toggle-history").on("click", function (e) {
			e.stopPropagation();
			const assetName = $(this).data("asset");
			const historyRow = $(`.history-row[data-asset="${assetName}"]`);
			const icon = $(this).find("i");

			historyRow.toggleClass("d-none");
			icon.toggleClass("fa-chevron-right fa-chevron-down");
		});
	}

	group_by_owner(assets) {
		const grouped = {};

		for (const asset of assets) {
			const ownerKey = asset.current_owner || "_no_owner";
			if (!grouped[ownerKey]) {
				grouped[ownerKey] = {
					owner_id: asset.current_owner,
					owner_name: asset.current_owner_name,
					assets: [],
					total_quantity: 0,
					total_value: 0,
				};
			}
			grouped[ownerKey].assets.push(asset);
			grouped[ownerKey].total_quantity += asset.quantity || 0;
			grouped[ownerKey].total_value += asset.total || 0;
		}

		// Сортування: спочатку з власником, потім без
		const sorted = {};
		const keys = Object.keys(grouped).sort((a, b) => {
			if (a === "_no_owner") return 1;
			if (b === "_no_owner") return -1;
			return (grouped[a].owner_name || "").localeCompare(grouped[b].owner_name || "");
		});

		for (const key of keys) {
			sorted[key] = grouped[key];
		}

		return sorted;
	}

	render_movement_history(history) {
		if (!history || history.length === 0) {
			return `<div class="text-muted">${__("Немає історії передачі")}</div>`;
		}

		return `
			<table class="table table-sm table-bordered mb-0">
				<thead>
					<tr>
						<th>${__("Дата")}</th>
						<th>${__("Тип руху")}</th>
						<th>${__("Звідки")}</th>
						<th>${__("Куди")}</th>
						<th>${__("Підстава")}</th>
					</tr>
				</thead>
				<tbody>
					${history
						.map(
							(m) => `
						<tr>
							<td>${frappe.datetime.str_to_user(m.date)}</td>
							<td>${m.movement_type}</td>
							<td>${m.from_source_name || "-"}</td>
							<td>${m.to_organization_name || "-"}</td>
							<td>
								${
									m.reference_appendix
										? `<a href="/app/oibasisdoc/${m.reference_appendix}">${
												m.reference_appendix_title || m.reference_appendix
										  }</a>`
										: "-"
								}
							</td>
						</tr>
					`
						)
						.join("")}
				</tbody>
			</table>
		`;
	}

	get_status_class(status) {
		const status_map = {
			"На складі": "badge-info",
			"В експлуатації": "badge-success",
			Передано: "badge-warning",
			"Очікує прийняття": "badge-secondary",
			Списується: "badge-danger",
			Списано: "badge-dark",
		};
		return status_map[status] || "badge-secondary";
	}
}

File: orange_inventory/orange_inventory/page/solution_assets/solution_assets.py
# Copyright (c) 2025, Maxim Sysoev and contributors
# For license information, please see license.txt

import frappe


@frappe.whitelist()
def get_assets_by_decision(decision):
	"""
	Отримує всі активи, пов'язані з рішенням (oiBasisDoc).
	Шукає активи через:
	1. oiReceiptOrder (прибуткові ордери) - активи, що надійшли за цим рішенням
	2. oiIssueOrder (видаткові ордери) - активи, що передавалися за цим рішенням
	3. oiAssetMovementHistory - історія передачі з посиланням на рішення
	"""
	if not decision:
		return {"assets": [], "decision_info": {}, "summary": {}}

	# Отримуємо інформацію про рішення
	decision_info = frappe.get_doc("oiBasisDoc", decision)

	# Знаходимо всі активи через історію руху
	asset_names = frappe.db.sql(
		"""
		SELECT DISTINCT parent
		FROM `taboiAssetMovementHistory`
		WHERE reference_appendix = %s
		AND parenttype = 'oiAsset'
	""",
		decision,
		as_dict=False,
	)
	asset_names = [a[0] for a in asset_names]

	# Шукаємо активи через видаткові ордери
	issue_orders = frappe.get_all(
		"oiIssueOrder", filters={"decision": decision, "docstatus": 1}, pluck="name"
	)

	for io in issue_orders:
		items = frappe.get_all("oiIssueOrderItem", filters={"parent": io}, pluck="asset")
		asset_names.extend(items)

	# Унікальні активи
	asset_names = list(set(asset_names))

	if not asset_names:
		return {
			"assets": [],
			"decision_info": {
				"decision_number": decision_info.decision_number,
				"decision_date": decision_info.decision_date,
				"title": decision_info.title,
				"full_title": decision_info.full_title,
			},
			"summary": {"total_quantity": 0, "total_value": 0, "status_counts": {}},
		}

	# Отримуємо дані активів
	assets = frappe.get_all(
		"oiAsset",
		filters={"name": ("in", asset_names)},
		fields=[
			"name",
			"asset_name",
			"serial_no",
			"inventory_no",
			"status",
			"current_owner",
			"quantity",
			"cost",
			"total",
		],
		order_by="asset_name",
	)

	# Отримуємо назви організацій
	org_names = {}
	org_list = list(set([a.current_owner for a in assets if a.current_owner]))
	if org_list:
		orgs = frappe.get_all(
			"oiOrganization", filters={"name": ("in", org_list)}, fields=["name", "organization_name"]
		)
		org_names = {o.name: o.organization_name for o in orgs}

	# Отримуємо історію руху для кожного активу
	for asset in assets:
		asset["current_owner_name"] = org_names.get(asset.current_owner, asset.current_owner)
		asset["movement_history"] = get_asset_movement_history(asset.name)

	# Підсумки
	summary = {
		"total_quantity": sum(a.quantity or 0 for a in assets),
		"total_value": sum(a.total or 0 for a in assets),
		"status_counts": {},
	}

	for asset in assets:
		status = asset.status or "Невідомо"
		summary["status_counts"][status] = summary["status_counts"].get(status, 0) + 1

	return {
		"assets": assets,
		"decision_info": {
			"decision_number": decision_info.decision_number,
			"decision_date": decision_info.decision_date,
			"title": decision_info.title,
			"full_title": decision_info.full_title,
		},
		"summary": summary,
	}


def get_asset_movement_history(asset_name):
	"""
	Отримує історію руху активу з розшифровкою посилань.
	"""
	history = frappe.get_all(
		"oiAssetMovementHistory",
		filters={"parent": asset_name, "parenttype": "oiAsset"},
		fields=[
			"date",
			"movement_type",
			"from_source_type",
			"from_source_name",
			"to_organization",
			"reference_appendix",
		],
		order_by="date desc",
	)

	# Збираємо унікальні організації та рішення для отримання назв
	org_names = set()
	decision_names = set()

	for h in history:
		if h.to_organization:
			org_names.add(h.to_organization)
		if h.from_source_type == "oiOrganization" and h.from_source_name:
			org_names.add(h.from_source_name)
		if h.reference_appendix:
			decision_names.add(h.reference_appendix)

	# Отримуємо назви організацій
	org_titles = {}
	if org_names:
		orgs = frappe.get_all(
			"oiOrganization",
			filters={"name": ("in", list(org_names))},
			fields=["name", "organization_name"],
		)
		org_titles = {o.name: o.organization_name for o in orgs}

	# Отримуємо назви донорів
	donor_names = set()
	for h in history:
		if h.from_source_type == "oiDonor" and h.from_source_name:
			donor_names.add(h.from_source_name)

	donor_titles = {}
	if donor_names:
		donors = frappe.get_all(
			"oiDonor", filters={"name": ("in", list(donor_names))}, fields=["name", "donor_name"]
		)
		donor_titles = {d.name: d.donor_name for d in donors}

	# Отримуємо назви рішень
	decision_titles = {}
	if decision_names:
		decisions = frappe.get_all(
			"oiBasisDoc",
			filters={"name": ("in", list(decision_names))},
			fields=["name", "full_title", "title"],
		)
		decision_titles = {d.name: d.full_title or d.title for d in decisions}

	# Додаємо назви до історії
	for h in history:
		h["to_organization_name"] = org_titles.get(h.to_organization, h.to_organization)
		if h.from_source_type == "oiOrganization":
			h["from_source_name"] = org_titles.get(h.from_source_name, h.from_source_name)
		elif h.from_source_type == "oiDonor":
			h["from_source_name"] = donor_titles.get(h.from_source_name, h.from_source_name)
		h["reference_appendix_title"] = decision_titles.get(h.reference_appendix, "")

	return history
